File: chapters/physics_solvers/verlet/code/c/verlet.c
// Submitted by Gathros
#include <stdio.h>

// Simple function for velocity-verlet
void verlet(double pos, double acc, double dt){

    // Note that we are using a temp variable for the previous position
    double prev_pos, temp_pos, time;
    prev_pos = pos;
    time = 0;

    while (pos > 0){
        time += dt;
        temp_pos = pos;
        pos = pos*2 - prev_pos + acc * dt * dt;
        prev_pos = temp_pos;
    }

   printf("%f\n", time);

}

// Simple function for stormer-verlet
void stormer_verlet(double pos, double acc, double dt){

    double prev_pos, temp_pos, time, vel;
    prev_pos = pos;
    vel = 0;
    time = 0;
    while (pos > 0){
        time += dt;
        temp_pos = pos;
        pos = pos*2 - prev_pos + acc * dt * dt;
        prev_pos = temp_pos;

        // The acceleration is constant, so the velocity is straightforward
        vel += acc*dt;
    }

   printf("%f\n", time);

}

// Simple function for velocity-verlet
void velocity_verlet(double pos, double acc, double dt){

    double time, vel;
    vel = 0;
    time = 0;
    while (pos > 0){
        time += dt;
        pos += vel*dt + 0.5*acc * dt * dt;
        vel += acc*dt;
    }

   printf("%f\n", time);

}

int main(){
    verlet(5.0, -10, 0.01);
    stormer_verlet(5.0, -10, 0.01);
    velocity_verlet(5.0, -10, 0.01);

}

File: chapters/physics_solvers/verlet/code/python2/verlet.py
# KingFredrickVI
class Ball:

    def __init__(self, pos=0, vel=0, acc=0):
        self.pos = pos
        self.vel = vel
        self.acc = acc


class Simulation:

    def __init__(self, obj, dt = 0.01):
        self.obj = obj
        self.dt = dt
        self.prev_pos = self.obj.pos

    def run(self):
        self.time = 0

        while self.obj.pos > 0:
            self.time += self.dt
            self.step()

    def step(self):
        pass


class Verlet(Simulation):
    
    def step(self):
        temp_pos = self.obj.pos
        self.obj.pos = self.obj.pos * 2 - self.prev_pos + self.obj.acc * self.dt * self.dt
        self.prev_pos = temp_pos

class Stormer_Verlet(Simulation):

    def step(self):
        temp_pos = self.obj.pos
        self.obj.pos = self.obj.pos * 2 - self.prev_pos + self.obj.acc * self.dt * self.dt
        self.prev_pos = temp_pos

        self.obj.vel += self.obj.acc * self.dt

class Velocity_Verlet(Simulation):

    def step(self):
        self.obj.pos += self.obj.vel * self.dt + 0.5 * self.obj.acc * self.dt * self.dt
        self.obj.vel += self.obj.acc * self.dt


def main():
    sim = Verlet(Ball(pos = 5.0, acc = -10))
    sim.run()

    print "Verlet:", sim.time

    sim = Stormer_Verlet(Ball(pos = 5.0, acc = -10))
    sim.run()

    print "Stormer Verlet:", sim.time

    sim = Velocity_Verlet(Ball(pos = 5.0, acc = -10))
    sim.run()

    print "Velocity Verlet:", sim.time


if __name__ == "__main__":
    main()



